# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'C:\\Users\\craig\\Documents\\Python Scripts\\Notebooks\\Guided_Capstone_Final'

In [3]:
path="C:\\Users\\craig\\Documents\\Python Scripts\\Notebooks\\Guided_Capstone_Final\\data"
os.chdir(path) 

In [4]:
df=pd.read_csv('step3_output.csv', index_col=0)

In [5]:
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple    ...     SkiableTerrain_ac  \
0          0          2     2       0    ...                1610.0   
1          0          0     0       0    ...                 640.0   
2          0          0     0       1    ...                  30.0   
3          1          0     2       2    ...                 777.0   
4          0          1     2       3    ...                 800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         1  
4          78.0              104.0            80.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [6]:
dfo=df['state'] # select object type columns
#dfo2=df.select_dtypes(include=['object'])

In [7]:
df = df.drop('state', axis=1)

In [8]:
df = pd.concat([df, pd.get_dummies(dfo)], axis=1)

In [9]:
print(df.columns)

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')


## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [10]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [11]:
X_scaled.std(axis = 0)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1.])

In [12]:
print(X_scaled)

[[-0.16898816  1.36939248  1.4766915  ... -0.11094004 -0.22609337
  -0.15786741]
 [-0.53057645  0.35051616 -0.30945373 ... -0.11094004 -0.22609337
  -0.15786741]
 [-0.66829865 -0.97190039 -0.30945373 ... -0.11094004 -0.22609337
  -0.15786741]
 ...
 [ 0.87581029  0.38341738 -0.30945373 ... -0.11094004 -0.22609337
   6.33442973]
 [ 1.37674105 -0.23321506 -0.30945373 ... -0.11094004 -0.22609337
   6.33442973]
 [ 1.33272395 -0.11646881 -0.30945373 ... -0.11094004 -0.22609337
   6.33442973]]


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [15]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled,y, test_size=0.25, random_state=1)

In [16]:
#No longer Need...

#np.delete(arr, 1, 0)
#np.set_printoptions(threshold=np.nan)
#y_test3 = np.delete(y_test2,20)
#print(y_test3)

In [24]:
print(y_train)

[ 45.          90.71428571  60.          54.          50.
  54.5         44.          56.          35.          32.
  52.57692308  93.          49.          69.          87.
  87.9         81.          50.          48.          65.
  49.          73.          79.          54.          47.
  43.          49.          71.          44.          20.
  44.          74.          40.          90.          76.
  42.          52.57692308  25.          52.          44.
  45.          45.          55.          50.          77.
  38.          89.53846154  59.          90.71428571  99.
  45.          74.          59.          68.          75.
  92.          54.26666667  89.53846154  88.          98.
  68.          91.         115.          90.71428571  90.71428571
  47.          89.          57.          60.          89.
  68.          90.71428571  90.71428571  56.          71.
  99.          17.          72.          35.          78.
  84.          41.          48.         119.          65.
  93. 

In [18]:
#No Longer Needed...

#np.set_printoptions(threshold=np.nan)
#X_test2 = np.delete(X_test,13,0)
#print(X_test2)

In [25]:
print(X_train)

[[-8.72181525e-01 -7.53266511e-01 -3.09453733e-01 -5.52157630e-02
  -2.84543897e-01 -4.64728957e-01 -7.13962162e-01 -3.12032280e-01
   9.04594951e-02  1.80487987e-01 -3.95650647e-01 -6.73621138e-01
  -6.61190043e-01 -2.20568533e-02 -4.28929581e-01  1.40341711e-01
  -1.28331797e-01 -1.71918862e-01 -6.15849079e-01 -5.40100487e-01
  -3.22014488e-01 -1.82672625e-01 -7.55320641e-01 -9.59294219e-02
  -7.82061887e-02 -2.61116484e-01 -2.61116484e-01 -1.24225999e-01
  -1.94563331e-01 -1.10940039e-01 -7.82061887e-02 -9.59294219e-02
  -1.67705098e-01 -5.52157630e-02 -1.85987083e-01  3.21633760e+00
  -2.10818511e-01 -7.82061887e-02 -1.94563331e-01 -1.10940039e-01
  -2.26093368e-01 -7.82061887e-02 -1.67705098e-01 -3.33895922e-01
  -1.36293255e-01 -1.24225999e-01 -1.77053558e-01 -2.47568824e-01
  -5.52157630e-02 -7.82061887e-02 -5.52157630e-02 -2.02828106e-01
  -2.18565095e-01 -1.10940039e-01 -1.77053558e-01 -1.10940039e-01
  -2.26093368e-01 -1.57867408e-01]
 [ 1.62113045e+00  2.60902533e+00  1.4766

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [26]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [27]:
print('Intercept: \n', model.intercept_)
print('Coefficients: \n', model.coef_)

Intercept: 
 18625327507.757397
Coefficients: 
 [ 4.49256781e+00 -4.38976435e-01  2.28597808e+10  3.24534530e+11
  2.66056267e+10  8.97289273e+10  5.35387778e+10  6.60239488e+10
  7.40853834e+10  8.38888927e+10 -2.36197898e+11  2.10433960e-01
  6.05468750e-02 -5.79902649e-01 -5.50153732e-01  2.56044006e+00
 -9.38980103e-01  2.57209778e-01  6.57226562e-01  2.06649399e+01
  7.46551514e-01  1.02068329e+00  8.09432983e-01 -1.29784880e+10
 -1.06131316e+10 -3.33764144e+10 -3.33764144e+10 -1.67036807e+10
 -2.55961666e+10 -1.49632644e+10 -1.06131316e+10 -1.29784880e+10
 -2.22715743e+10 -7.51608348e+09 -2.45450870e+10 -3.87092093e+10
 -2.75596672e+10 -1.06131316e+10 -2.55961666e+10 -1.49632644e+10
 -2.93688427e+10 -1.03802831e+04 -2.22715743e+10 -4.10163961e+10
 -1.82697060e+10 -1.67036807e+10 -2.34395902e+10 -3.18502115e+10
 -7.51608348e+09 -1.06131316e+10 -7.51608348e+09 -2.65992806e+10
 -2.84816488e+10 -1.49632644e+10 -2.34395902e+10 -1.49632644e+10
 -2.93688427e+10 -2.10306252e+10]


**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [28]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test)

In [29]:
print(y_pred)

[5.85535393e+01 5.94211082e+01 1.21138988e+02 8.61287613e+01
 5.46647453e+01 6.13007011e+01 7.99584923e+01 6.14243584e+01
 5.60691719e+01 8.50531731e+01 7.95770378e+01 3.68328514e+01
 3.90618095e+01 6.83946800e+01 4.52051201e+01 5.46070595e+01
 6.26803398e+01 8.87510338e+01 7.45606041e+01 9.43897743e+01
 7.33909569e+01 5.42237663e+01 2.82176933e+01 6.46322250e+01
 9.86568108e+01 7.36455421e+01 1.18750866e+02 5.56565208e+01
 7.11933861e+01 8.49343834e+01 1.25061398e+02 7.33644295e+01
 3.89393311e+01 6.86046410e+01 3.70342522e+01 5.36284409e+01
 6.52870903e+01 5.20589561e+01 1.36536934e+11 7.84583244e+01
 5.85465886e+12 8.40744171e+01 1.70043530e+02 7.69345055e+01
 5.80625725e+01 7.85602074e+01 9.01260490e+01 5.97759209e+01
 2.29028816e+01 3.86963463e+01 1.12826267e+02 6.02710991e+01
 6.36619987e+01 6.21569633e+01 2.66919670e+01 8.99121819e+01
 5.11042023e+01 3.82467651e+01 4.46995354e+01 3.58037376e+01
 5.96639824e+01 9.39886742e+01 7.89416008e+01 3.39949760e+01
 7.42391167e+01 5.312295

## Review Model Outcomes — Iterate over additional models as needed

In [0]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [30]:
vs = explained_variance_score(y_test, y_pred)

In [31]:
print(vs)

-6.001903445661466e+20


In [32]:
ae = mean_absolute_error(y_test, y_pred)

In [33]:
print(ae)

73828105168.15376


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [144]:
print(model.intercept_)

73032336319255.56


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [34]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefdf= pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


In [35]:
coefdf.sort_values(by='Coefficient', ascending=False)

Coefficient
fastEight          3.245345e+11
total_chairs       2.361979e+11
fastQuads          8.972893e+10
surface            8.388889e+10
double             7.408538e+10
triple             6.602395e+10
quad               5.353878e+10
New York           4.101640e+10
Michigan           3.870921e+10
Colorado           3.337641e+10
California         3.337641e+10
Pennsylvania       3.185021e+10
Wisconsin          2.936884e+10
New Hampshire      2.936884e+10
Vermont            2.848165e+10
Minnesota          2.755967e+10
fastSixes          2.660563e+10
Utah               2.659928e+10
Montana            2.559617e+10
Idaho              2.559617e+10
Massachusetts      2.454509e+10
Oregon             2.343959e+10
Washington         2.343959e+10
trams              2.285978e+10
New Mexico         2.227157e+10
Maine              2.227157e+10
Wyoming            2.103063e+10
North Carolina     1.826971e+10
Ohio               1.670368e+10
Connecticut        1.670368e+10
Nevada             1.496326e+10
Illinois           1.496326e+10
West Virginia      1.496326e+10
Virginia           1.496326e+10
Alaska             1.297849e+10
Iowa               1.297849e+10
South Dakota       1.061313e+10
Arizona            1.061313e+10
Missouri           1.061313e+10
Indiana            1.061313e+10
Rhode Island       7.516083e+09
Maryland           7.516083e+09
Tennessee          7.516083e+09
New Jersey         1.038028e+04
AdultWeekday       2.066494e+01
summit_elev        4.492568e+00
Snow Making_ac     2.560440e+00
NightSkiing_ac     1.020683e+00
daysOpenLastYear   9.389801e-01
clusters           8.094330e-01
projectedDaysOpen  7.465515e-01
averageSnowfall    6.572266e-01
LongestRun_mi      5.799026e-01
SkiableTerrain_ac  5.501537e-01
vertical_drop      4.389764e-01
yearsOpen          2.572098e-01
Runs               2.104340e-01
TerrainParks       6.054688e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [36]:
df2=pd.read_csv('step3_output.csv', index_col=0)

In [37]:
df2 = df2.drop('state', axis=1)

In [38]:
X2 = df2.drop(['Name','AdultWeekend'], axis=1)
y2 = df2.AdultWeekend
scaler2 = preprocessing.StandardScaler().fit(X2)
X_scaled2=scaler2.transform(X2) 

In [39]:
y2 = y2.ravel()
X2_train, X2_test, y2_train, y2_test = train_test_split(X_scaled2,y2, test_size=0.25, random_state=1)

In [40]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X2_train,y2_train)

In [41]:
y2_pred = model2.predict(X2_test)

In [42]:
print(y2_pred)

[ 58.59139789  53.16019523 116.87718704  95.55888936  56.48995952
  51.15978147  82.24024255  66.06530353  64.13140519  87.75054116
  79.71824778  44.30244132  39.03175236  73.80234026  36.56842816
  56.45804134  54.24419487  84.76650855  75.84544995  87.58545179
  72.39572964  49.96556044  28.49582998  64.7581076   91.65561358
  65.41925757 111.19746209  48.44363788  72.8588865   76.28858399
 133.51687771  82.08957003  30.30255821  70.77030238  36.72120913
  47.01132027  72.17456782  52.50055847  85.50725974  70.66261482
  56.42176565  85.81146115 167.48813446  72.16619436  61.35514376
  75.38957206  96.14880406  60.50868567  24.42639184  47.10931434
 105.27229479  65.48043153  57.16908734  58.34060772  33.56714314
  87.3597109   54.06528574  41.14362265  38.88503858  36.54076587
  59.73634946  98.05188271  71.39915421  46.11937208  79.80648776
  54.51310036 149.04551472  60.27015301  43.77038332  51.58398358
  48.06275455  77.1354825   86.52602448  89.54955984  74.34563392
  93.42591

In [43]:
print(model2.intercept_)

65.59171020713525


In [44]:
vs2 = explained_variance_score(y2_test, y2_pred)

In [45]:
print(vs2)

0.9176001177884492


In [46]:
print(mean_absolute_error(y2_test, y2_pred))

5.606589114282675


In [47]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefdf2= pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])


In [48]:
coefdf2.sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday       2.213472e+01
Snow Making_ac     2.162400e+00
vertical_drop      1.348725e+00
SkiableTerrain_ac  1.240977e+00
NightSkiing_ac     1.167829e+00
quad               1.069275e+00
trams              1.023326e+00
fastSixes          1.016621e+00
surface            9.422165e-01
daysOpenLastYear   9.043981e-01
clusters           7.980488e-01
averageSnowfall    7.307063e-01
summit_elev        6.363575e-01
triple             6.070318e-01
projectedDaysOpen  5.796903e-01
fastQuads          5.405734e-01
TerrainParks       4.927967e-01
LongestRun_mi      3.908362e-01
total_chairs       3.829536e-01
Runs               2.956389e-01
double             2.700688e-01
yearsOpen          2.121101e-01
fastEight          1.776357e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [49]:
X3 = df2.drop(['Name','AdultWeekend', 'summit_elev'], axis=1)
y3 = df2.AdultWeekend
scaler3 = preprocessing.StandardScaler().fit(X3)
X_scaled3=scaler3.transform(X3) 

In [50]:
y3 = y3.ravel()
X3_train, X3_test, y3_train, y3_test = train_test_split(X_scaled3,y3, test_size=0.25, random_state=1)

In [51]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train,y3_train)

In [52]:
y3_pred = model3.predict(X3_test)

In [53]:
print(y3_pred)

[ 58.6922296   53.09022421 116.74969183  95.37433018  56.20340755
  51.80028444  83.05935338  66.04634098  63.99678628  87.43976343
  80.43080998  44.22794193  38.85251091  74.05397768  36.53486818
  56.13264438  54.0285783   85.20774742  75.57580313  88.11552009
  72.36567225  49.73619696  28.49133569  64.72099967  91.92247181
  65.69825938 111.2569793   48.26369189  72.74566178  76.43915147
 133.65022587  81.36037773  30.35002421  70.39070484  36.62858742
  46.78723298  71.70771522  52.51315655  85.50245552  70.90963486
  55.93255841  86.18714829 167.42670549  71.95470579  61.65921361
  75.42773447  96.36336678  60.26442937  24.33593037  47.2222316
 103.98959376  65.32640223  57.47268083  58.22305534  33.53981033
  86.38257064  55.01422383  40.90820682  38.93806134  36.44521007
  59.71981817  97.42854841  71.33476556  46.51799183  80.38564956
  54.2854494  149.03898605  60.24136956  43.68670489  51.4693909
  48.11522906  76.8986085   86.39771946  89.76079423  74.22392478
  93.0674484

In [54]:
print(model3.intercept_)

65.5823220324717


In [55]:
print(mean_absolute_error(y3_test, y3_pred))

5.642964316390303


In [56]:
print(explained_variance_score(y3_test, y3_pred))

0.9172978430861458


In [57]:
coefdf3= pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])

In [58]:
coefdf3.sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday       2.204980e+01
Snow Making_ac     2.146234e+00
SkiableTerrain_ac  1.277429e+00
NightSkiing_ac     1.230322e+00
vertical_drop      1.226097e+00
quad               1.109385e+00
clusters           1.094877e+00
fastSixes          1.009854e+00
surface            9.749420e-01
trams              9.585177e-01
averageSnowfall    8.286131e-01
daysOpenLastYear   7.707339e-01
triple             5.968123e-01
fastQuads          5.915181e-01
Runs               4.698999e-01
projectedDaysOpen  4.575351e-01
TerrainParks       4.547720e-01
total_chairs       4.097023e-01
LongestRun_mi      3.903166e-01
double             2.932566e-01
yearsOpen          2.217173e-01
fastEight          1.443290e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -6.00e-20 | 7.3e11|-|
| Model 2. | 0.917 | 5.60 |'state'|
| Model 3. | 0.917 | 5.64 |'state','summit_elev','base_elev'|

Model Selection: Model 3 or 2